File: 2021/python/day5.py
infile = open('inputs/day5.in', 'r')
lines = infile.readlines()

grid = {}

for line in lines:
    start, end = line.split(' -> ')
    x1, y1 = map(int, start.split(','))
    x2, y2 = map(int, end.split(','))

    if x1 != x2 and y1 != y2:
        continue

    if x1 == x2:
        if y1 > y2:
            x1, y1, x2, y2 = x2, y2, x1, y1

        for y in range(y1, y2 + 1):
            grid.setdefault((x1, y), 0)
            grid[(x1, y)] += 1

    if y1 == y2:
        if x1 > x2:
            x1, y1, x2, y2 = x2, y2, x1, y1

        for x in range(x1, x2 + 1):
            grid.setdefault((x, y1), 0)
            grid[(x, y1)] += 1

res = sum(1 for k, v in grid.items() if v >= 2)
print(res)

# Diagonal lines
for line in lines:
    start, end = line.split(' -> ')
    x1, y1 = map(int, start.split(','))
    x2, y2 = map(int, end.split(','))

    if y2 - y1 == x2 - x1:
        if x1 > x2:
            x1, y1, x2, y2 = x2, y2, x1, y1

        for i in range(x2 - x1 + 1):
            grid.setdefault((x1 + i, y1 + i), 0)
            grid[(x1 + i, y1 + i)] += 1

    if y2 - y1 == x1 - x2:
        if x1 > x2:
            x1, y1, x2, y2 = x2, y2, x1, y1

        for i in range(x2 - x1 + 1):
            grid.setdefault((x1 + i, y1 - i), 0)
            grid[(x1 + i, y1 - i)] += 1

res = sum(1 for k, v in grid.items() if v >= 2)
print(res)

File: 2021/elixir/day5.ex
{:ok, lines} = File.read("inputs/day5.in")

lines =
  lines
  |> String.split("\r\n", trim: true)
  |> Enum.map(fn line ->
    line
    |> String.split(" -> ", trim: true)
    |> Enum.map(&String.split(&1, ",", trim: true))
    |> Enum.reduce(&(&2 ++ &1))
    |> Enum.map(&String.to_integer/1)
  end)

h_and_v_lines =
  Enum.filter(lines, fn [x1, y1, x2, y2] ->
    x1 == x2 || y1 == y2
  end)

IO.inspect(h_and_v_lines)

line_map =
  Enum.reduce(h_and_v_lines, %{}, fn [x1, y1, x2, y2], acc ->
    if y1 == y2 do
      if x1 > x2 do
        [x1, y1, x2, y2] = [x2, y2, x1, y1]
      end

      for x <- x1..x2,
          do:
            Map.get_and_update(acc, {x, y1}, fn current ->
              {current, 1}
            end)
    else
      if y1 > y2 do
        [x1, y1, x2, y2] = [x2, y2, x1, y1]
      end

      for y <- y1..y2,
          do:
            Map.get_and_update(acc, {x1, y}, fn current ->
              {current, 1}
            end)
    end
  end)

IO.inspect(Map.get(line_map, {0, 9}))
